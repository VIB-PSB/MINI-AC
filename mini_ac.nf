nextflow.enable.dsl = 2

include { genome_wide_miniac } from './workflows/miniac_gw'
include { locus_based_miniac } from './workflows/miniac_lb'

workflow MINIAC {

    params.OBO_file = "$projectDir/data/ontologies/go.obo"

    // define species id used for data subfolder and data file prefix
    def species
    switch(params.species) {
        case "arabidopsis":
            species = "ath"
            break
        case "maize_v4":
            species = "zma_v4"
            break
        case "maize_v5":
            species = "zma_v5"
            break
        default:
            exit 1, "MINI-AC can only be run for the species 'arabidopsis', 'maize_v4' and 'maize_v5'. Instead it got '${params.species}'."
    }

    // set input data parameters shared between genome-wide and locus-based modes
    params.Faix_file = "$projectDir/data/${species}/${species}.fasta.fai"
    params.Motif_tf_file = "$projectDir/data/${species}/${species}_motif_TF_file.txt"
    params.Genes_coords = "$projectDir/data/${species}/${species}_genes_coords_sorted.bed"
    params.Feature_file = "$projectDir/data/${species}/${species}_go_gene_file.txt"
    params.TF_fam_file = "$projectDir/data/${species}/${species}_TF_family_file.txt"
    params.Genes_metadata = "$projectDir/data/${species}/${species}_gene_metadata_file.txt"

    // set defaut p-value
    params.P_val = 0.1

    if (params.mode == "genome_wide") {
        
        params.MotMapsFile = "$projectDir/data/${species}/${species}_genome_wide_motif_mappings.bed"
        params.Non_cod_genome = "$projectDir/data/${species}/${species}_noncod_merged.bed"

        genome_wide_miniac(params.OutDir, params.ACR_dir, params.Filter_set_genes, params.Set_genes_dir, params.One_filtering_set, params.DE_genes, params.DE_genes_dir, params.One_DE_set, params.P_val, params.Bps_intersect, params.Second_gene_annot, params.Second_gene_dist, params.MotMapsFile, params.Non_cod_genome, params.Faix_file, params.Motif_tf_file, params.Genes_coords, params.Feature_file, params.OBO_file, params.TF_fam_file, params.Genes_metadata)
    
    } else if (params.mode == "locus_based") {

        params.MotMapsFile = "$projectDir/data/${species}/${species}_locus_based_motif_mappings_5kbup_1kbdown.bed"
        params.Promoter_file = "$projectDir/data/${species}/${species}_promoter_5kbup_1kbdown_sorted.bed"
        
        locus_based_miniac(params.OutDir, params.ACR_dir, params.Filter_set_genes, params.Set_genes_dir, params.One_filtering_set, params.DE_genes, params.DE_genes_dir, params.One_DE_set, params.P_val, params.Bps_intersect, params.MotMapsFile, params.Promoter_file, params.Faix_file, params.Motif_tf_file, params.Feature_file, params.OBO_file, params.TF_fam_file, params.Genes_metadata)
    
    } else {
        exit 1, "MINI-AC can only be run using the modes 'genome_wide' or 'locus_based'. Instead it got '${params.mode}'."
    }
}


workflow {
    MINIAC()
}