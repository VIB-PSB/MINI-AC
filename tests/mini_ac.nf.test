nextflow_workflow {

    name "Test Workflow MINIAC"
    script "mini_ac.nf"
    workflow "MINIAC"

    
    test("maize_v4 genome_wide") {

        when {
            params {
                //// Mode
                mode = "genome_wide"

                //// Species
                species = "maize_v4"

                //// Lower shuffle count to reduce test runtime
                Shuffle_count = 100

                //// Fix shuffle seed for reproducible test results
                Shuffle_seed = 42

                //// Hard code data paths
                MotMapsFile_gw = "${baseDir}/tests/data/zma_v4/zma_v4_genome_wide_motif_mappings_chr1.bed"
                Non_cod_genome = "${baseDir}/tests/data/zma_v4/zma_v4_noncod_merged_chr1.bed"
                Faix_file = "${baseDir}/data/zma_v4/zma_v4.fasta.fai"
                Motif_tf_file = "${baseDir}/data/zma_v4/zma_v4_motif_TF_file.txt"
                Genes_coords = "${baseDir}/data/zma_v4/zma_v4_genes_coords_sorted.bed"
                Feature_file = "${baseDir}/data/zma_v4/zma_v4_go_gene_file.txt"
                TF_fam_file = "${baseDir}/data/zma_v4/zma_v4_TF_family_file.txt"
                Genes_metadata = "${baseDir}/data/zma_v4/maize_v4_gene_metadata_file.txt"
                OBO_file = "${baseDir}/data/ontologies/go.obo"

                //// Output folder
                OutDir = "${baseDir}/tests/outputs/zma_v4/genome_wide"

                //// Output CSV instead of Excel for automated output comparison
                Csv_output = true

                //// Required input
                ACR_dir = "${baseDir}/tests/inputs/zma_v4/acr_files"

                //// Optional input
                // Differential expression data
                DE_genes = true
                DE_genes_dir = "${baseDir}/tests/inputs/zma_v4/de_files"
                One_DE_set = true
                // Expression data
                Filter_set_genes = true
                Set_genes_dir = "${baseDir}/tests/inputs/zma_v4/exp_genes_files"
                One_filtering_set = true

                //// Prediction parameters
                Bps_intersect = false

                //// Prediction parameters only genome-wide
                Second_gene_annot = true
                Second_gene_dist = 500
            }
            workflow {
                """
                // define inputs of the workflow here. Example:
                // input[0] = file("test-file.txt")
                """
            }
        }

        then {
            assert workflow.success
            assert snapshot(
                workflow,
                path("${params.OutDir}/enrichment_stats").list(),
                path("${params.OutDir}/GO_enrichment").list(),
                path("${params.OutDir}/integrative_outputs").list(),
                path("${params.OutDir}/networks").list(),
            ).match()
        }
    }

    test("maize_v4 locus_based") {

        when {
            params {
                //// Mode
                mode = "locus_based"

                //// Species
                species = "maize_v4"

                //// Lower shuffle count to reduce test runtime
                Shuffle_count = 200

                //// Fix shuffle seed for reproducible test results
                Shuffle_seed = 42

                //// Hard code data paths
                MotMapsFile_lb = "${baseDir}/tests/data/zma_v4/zma_v4_locus_based_motif_mappings_5kbup_1kbdown_chr1.bed"
                Promoter_file = "${baseDir}/tests/data/zma_v4/zma_v4_promoter_5kbup_1kbdown_sorted_chr1.bed"
                Faix_file = "${baseDir}/data/zma_v4/zma_v4.fasta.fai"
                Motif_tf_file = "${baseDir}/data/zma_v4/zma_v4_motif_TF_file.txt"
                Feature_file = "${baseDir}/data/zma_v4/zma_v4_go_gene_file.txt"
                TF_fam_file = "${baseDir}/data/zma_v4/zma_v4_TF_family_file.txt"
                Genes_metadata = "${baseDir}/data/zma_v4/maize_v4_gene_metadata_file.txt"
                OBO_file = "${baseDir}/data/ontologies/go.obo"

                //// Output folder
                OutDir = "${baseDir}/tests/outputs/zma_v4/locus_based"

                //// Output CSV instead of Excel for automated output comparison
                Csv_output = true

                //// Required input
                ACR_dir = "${baseDir}/tests/inputs/zma_v4/acr_files"

                //// Optional input
                // Differential expression data
                DE_genes = false
                // Expression data
                Filter_set_genes = true
                Set_genes_dir = "${baseDir}/tests/inputs/zma_v4/exp_genes_files"
                One_filtering_set = true

                //// Prediction parameters
                Bps_intersect = true
            }
            workflow {
                """
                // define inputs of the workflow here. Example:
                // input[0] = file("test-file.txt")
                """
            }
        }

        then {
            assert workflow.success
            assert snapshot(
                workflow,
                path("${params.OutDir}/enrichment_stats").list(),
                path("${params.OutDir}/GO_enrichment").list(),
                path("${params.OutDir}/integrative_outputs").list(),
                path("${params.OutDir}/networks").list(),
            ).match()
        }
    }

    test("arabidopsis genome_wide") {

        when {
            params {
                //// Mode
                mode = "genome_wide"

                //// Species
                species = "arabidopsis"

                //// Lower shuffle count to reduce test runtime
                Shuffle_count = 100

                //// Fix shuffle seed for reproducible test results
                Shuffle_seed = 42

                //// Hard code data paths
                MotMapsFile_gw = "${baseDir}/tests/data/ath/ath_genome_wide_motif_mappings.bed"
                Non_cod_genome = "${baseDir}/data/ath/ath_noncod_merged.bed"
                Faix_file = "${baseDir}/data/ath/ath.fasta.fai"
                Motif_tf_file = "${baseDir}/data/ath/ath_motif_TF_file.txt"
                Genes_coords = "${baseDir}/data/ath/ath_genes_coords_sorted.bed"
                Feature_file = "${baseDir}/data/ath/ath_go_gene_file.txt"
                TF_fam_file = "${baseDir}/data/ath/ath_TF_family_file.txt"
                Genes_metadata = "${baseDir}/data/ath/arabidopsis_gene_metadata_file.txt"
                OBO_file = "${baseDir}/data/ontologies/go.obo"

                //// Output folder
                OutDir = "${baseDir}/tests/outputs/ath/genome_wide"

                //// Output CSV instead of Excel for automated output comparison
                Csv_output = true

                //// Required input
                ACR_dir = "${baseDir}/tests/inputs/ath/acr_files"

                //// Optional input
                // Differential expression data
                DE_genes = true
                DE_genes_dir = "${baseDir}/tests/inputs/ath/de_files"
                One_DE_set = false
                // Expression data
                Filter_set_genes = false

                //// Prediction parameters
                Bps_intersect = true

                //// Prediction parameters only genome-wide
                Second_gene_annot = false
            }
            workflow {
                """
                // define inputs of the workflow here. Example:
                // input[0] = file("test-file.txt")
                """
            }
        }

        then {
            assert workflow.success
            assert snapshot(
                workflow,
                path("${params.OutDir}/enrichment_stats").list(),
                path("${params.OutDir}/GO_enrichment").list(),
                path("${params.OutDir}/integrative_outputs").list(),
                path("${params.OutDir}/networks").list(),
            ).match()
        }
    }

    test("arabidopsis locus_based") {

        when {
            params {
                //// Mode
                mode = "locus_based"

                //// Species
                species = "arabidopsis"

                //// Lower shuffle count to reduce test runtime
                Shuffle_count = 200

                //// Fix shuffle seed for reproducible test results
                Shuffle_seed = 42

                //// Hard code data paths
                MotMapsFile_lb = "${baseDir}/tests/data/ath/ath_locus_based_motif_mappings_5kbup_1kbdown.bed"
                Promoter_file = "${baseDir}/data/ath/ath_promoter_5kbup_1kbdown_sorted.bed"
                Faix_file = "${baseDir}/data/ath/ath.fasta.fai"
                Motif_tf_file = "${baseDir}/data/ath/ath_motif_TF_file.txt"
                Feature_file = "${baseDir}/data/ath/ath_go_gene_file.txt"
                TF_fam_file = "${baseDir}/data/ath/ath_TF_family_file.txt"
                Genes_metadata = "${baseDir}/data/ath/arabidopsis_gene_metadata_file.txt"
                OBO_file = "${baseDir}/data/ontologies/go.obo"

                //// Output folder
                OutDir = "${baseDir}/tests/outputs/ath/locus_based"

                //// Output CSV instead of Excel for automated output comparison
                Csv_output = true

                //// Required input
                ACR_dir = "${baseDir}/tests/inputs/ath/acr_files"

                //// Optional input
                // Differential expression data
                DE_genes = false
                // Expression data
                Filter_set_genes = false

                //// Prediction parameters
                Bps_intersect = false
            }
            workflow {
                """
                // define inputs of the workflow here. Example:
                // input[0] = file("test-file.txt")
                """
            }
        }

        then {
            assert workflow.success
            assert snapshot(
                workflow,
                path("${params.OutDir}/enrichment_stats").list(),
                path("${params.OutDir}/GO_enrichment").list(),
                path("${params.OutDir}/integrative_outputs").list(),
                path("${params.OutDir}/networks").list(),
            ).match()
        }
    }
}
