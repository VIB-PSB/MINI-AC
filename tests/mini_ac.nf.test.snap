{
    "maize_v4 genome_wide": {
        "content": [
            {
                "stderr": [
                    "Picked up JAVA_TOOL_OPTIONS: -XX:ParallelGCThreads=1"
                ],
                "errorReport": null,
                "exitStatus": 0,
                "out": {
                    
                },
                "failed": false,
                "stdout": [
                    
                ],
                "errorMessage": null,
                "trace": {
                    "tasksFailed": 0,
                    "tasksCount": 12,
                    "tasksSucceeded": 12
                },
                "name": "workflow",
                "success": true
            },
            [
                "bundle_sheath_marand_top10k_chr1_allshuff_sorted_miniac_stats.txt:md5,1e9f5e9dd74a7d87f70d457560ef5582",
                "mesophyll_marand_top10k_chr1_allshuff_sorted_miniac_stats.txt:md5,c0b46c8893eb2de41023235157e7b545"
            ],
            [
                "bundle_sheath_marand_top10k_chr1_GO_enrichment.txt:md5,147f8825d5cc12c28d506b30ec4c754a",
                "mesophyll_marand_top10k_chr1_GO_enrichment.txt:md5,3b4e2bdcbe2b8a902a22bee40f7efc50"
            ],
            [
                "bundle_sheath_marand_top10k_chr1_GO_enrichment.csv:md5,b7c00d0029349a07d24b9c8e09b487d0",
                "bundle_sheath_marand_top10k_chr1_TF_centric.csv:md5,3b5e27cb8137427fb0a3064ee8e8f3bc",
                "bundle_sheath_marand_top10k_chr1_functional_network.txt:md5,4bb035ce16e2cc08de102e5f1a1e50f0",
                "bundle_sheath_marand_top10k_chr1_motif_centric.csv:md5,e243ebc905669c6bc370017f570a1226",
                "bundle_sheath_marand_top10k_chr1_node_attributes.txt:md5,9a8695f13204877ef3ba00f3dcde0d49",
                "mesophyll_marand_top10k_chr1_GO_enrichment.csv:md5,e0df3df69365e1c9aeec86655ae56320",
                "mesophyll_marand_top10k_chr1_TF_centric.csv:md5,8c9ff0fa88ae3df64cab220ae2ce1d84",
                "mesophyll_marand_top10k_chr1_functional_network.txt:md5,bbd0e70667415779799281749cf5dbc6",
                "mesophyll_marand_top10k_chr1_motif_centric.csv:md5,33f411d3d31c1ff9e266740911f50e49",
                "mesophyll_marand_top10k_chr1_node_attributes.txt:md5,ab4edb60d8ae6dfad60ce592d98e9fb8"
            ],
            [
                "bundle_sheath_marand_top10k_chr1_network.txt:md5,216ceb10b36d17b10e68f9bf5256d8eb",
                "bundle_sheath_marand_top10k_chr1_network_filtered.txt:md5,abf45ebb46c870e023cf8cde900e914e",
                "mesophyll_marand_top10k_chr1_network.txt:md5,eda5ca6e1146b60b5df56adc11e33781",
                "mesophyll_marand_top10k_chr1_network_filtered.txt:md5,2f46652db211543d395bfb8662ec1f31"
            ]
        ],
        "timestamp": "2023-09-26T10:05:01+0000"
    }
}